### PR TITLE
Added CombineConsecutiveUnsetsFixer

### PR DESCRIPTION
**Pull request by SpacePossum**
Expected:
```php
    <?php unset($y, $u); ?>
```
Input:
```php
    <?php unset($y);unset($u) ?>
````

### CODE REVIEW

File: Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+class MultipleUnsetFixer  extends AbstractFixer
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function fix(\SplFileInfo $file, $content)
+    {
+        $tokens = Tokens::fromCode($content);
+
+        // Lower $count  with 6, because any call to 'unset' that can possible be merged is build up out of at least 5 Tokens; 'unset', '(', '[var]', ')', ';'.
```
Comment (GrahamCampbell) [inline Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php]: comment too long, and too many spaces before the word `with`

File: Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            }
+
+            // Remove 'unset', '(', ')' and ';' from the merged 'unset' call.
+            $tokens[$nextUnsetIndex + $contentTokenCount]->clear();
+            $tokens[$nextUnsetContentStart + $contentTokenCount]->clear();
+            $tokens[$nextUnsetContentEnd + $contentTokenCount]->clear();
+            $nextUnsetSemiColonIndex = $tokens->getNextTokenOfKind($nextUnsetContentEnd + $contentTokenCount, array(';'));
+            $tokens[$nextUnsetSemiColonIndex]->clear();
+
+            // Since tokens are cleared there might be successive whitespace tokens which need to be merge into one whitespace token.
```
Comment (GrahamCampbell) [inline Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php]: comment too long

File: Symfony/CS/Tests/Fixer/Contrib/MultipleUnsetFixerTest.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                '<?php //2
+                    unset($foo,$bar,$foobar,$foobar2);/*1*/
+                    /*2*/
+                    //3
+                    /*4*/
+                ',
+                '<?php //2
+                    unset($foo);/*1*/
+                    unset($bar);/*2*/
+                    unset($foobar);//3
```
Comment (GrahamCampbell) [inline Symfony/CS/Tests/Fixer/Contrib/MultipleUnsetFixerTest.php]: we're missing a test case for when comments appear on their own line, or the unsets are separated by an extra line of whitespace

File: Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        $tokens = Tokens::fromCode($content);
+
+        // Lower $count  with 6, because any call to 'unset' that can possible be merged is build up out of at least 5 Tokens; 'unset', '(', '[var]', ')', ';'.
+        for ($index = 0, $count = $tokens->count(); $index < $count - 6; ++$index) {
+
+            if (!$tokens[$index]->isGivenKind(T_UNSET)) {
+                continue;
+            }
+
+            $unsetEndSemicolon = $tokens->getNextTokenOfKind($index, array(';'));
```
Comment (gharlan) [inline Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php]: this is valid (without semicolon):

```php
<?php unset($x) ?>
```

```

File: Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            }
+
+            if (!$tokens[$nextUnsetIndex]->isGivenKind(T_UNSET)) {
+                $index = $nextUnsetIndex;
+                continue;
+            }
+
+            // Merge the tokens inside the 'unset' call into the previous one 'unset' call.
+            $unsetEndParenthesisBrace = $tokens->getPrevTokenOfKind($unsetEndSemicolon, array(')'));
+            $tokens->insertAt($unsetEndParenthesisBrace, array(new Token(',')));
```
Comment (gharlan) [inline Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php]: no need for the array

File: Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php (hunk 4 above)
Comment (SpacePossum) [inline Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php]: Thanks for the pointer, I'll update the fixer.
I think the `IncludeFixer` might need a fix for this as well (at least it is not in the utest of the fixer):
```php
<?php include('a') ?>
```

Comment (GrahamCampbell) [inline Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php]: How much common code is there between the two fixers?

Comment (SpacePossum) [inline Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php]: almost nothing

**Comment by SpacePossum** [comment]
Thanks for the review so far all! I'll rework the fixer; check the Tokens from last to first, making it more easy, split it up a bit in some private methods, fix the case were `unset` is used without a semicolon.
I'll comment if it is ready for re-review :)

**Comment by SpacePossum** [comment]
Refactored.

Todo; Priority + testing;

Should ran before:
* `trailing_spaces`
* `whitespacy_lines`
* `extra_empty_lines`

Should ran after:
* `duplicate_semicolon`

File: Symfony/CS/Tests/Fixer/Contrib/MultipleUnsetFixerTest.php (hunk 3 above)
Comment (SpacePossum) [inline Symfony/CS/Tests/Fixer/Contrib/MultipleUnsetFixerTest.php]: :+1:  let me know if there are still tests missing

**Comment by SpacePossum** [comment]
all clean upped, integration tests added

File: Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+                $tokens[$nextUnsetSemicolon]->clear();
+                $highestClearTokenIndex = $nextUnsetSemicolon;
+            } else {
+                $highestClearTokenIndex = $tokensAddCount + $nextUnsetContentEnd;
+            }
+
+            // Since tokens are cleared and inserted there might be successive whitespace tokens which need to
+            // be merge into one whitespace token. These tokens can only be after the first 'unset' call ';' token
+            // up to the token after the last cleared token of the second 'unset' + 1
+            $this->mergeWhitespaceTokens(
```
Comment (keradus) [inline Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php]: It is  the 2nd (+?) place of that kind of logic, we should extract it to one place.

File: README.rst [hunk 7]
```diff
@@ -634,6 +634,11 @@ Choose from the list of available fixers:
                         closing semicolon are
                         prohibited.
 
+* **multiple_unset** [contrib]
```
Comment (keradus) [inline README.rst]: btw, have you checked Symfony ?

Comment (SpacePossum) [inline README.rst]: I'm in the process, found a bug right away :(

Comment (SpacePossum) [inline README.rst]: here we are https://github.com/symfony/symfony/pull/17356 :)

Comment (keradus) [inline README.rst]: > I'm in the process, found a bug right away :(

yep, it;s always cool to test fixer on bigger codebase ;)

File: Symfony/CS/Tests/Fixer/Contrib/MultipleUnsetFixerTest.php [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+            ),
+            array(
+                '<?php unset($x,$b  , $d);  /**/   ?> b',
+                '<?php unset($x);  /**/ unset ($b  , $d) ?> b',
+            ),
+            array(
+                '<?php unset($x) ?>',
+            ),
+            array(
+                '<?php unset($y,$u); ?>',
```
Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/MultipleUnsetFixerTest.php]: the simplest case and it is on the end... but found it!
why without the space? we add a comma, we should add a space.

File: Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php (hunk 6 above)
Comment (SpacePossum) [inline Symfony/CS/Fixer/Contrib/MultipleUnsetFixer.php]: it is .. .clearing tokens seems tricky... let me think about it

File: README.rst (hunk 7 above)
Comment (SpacePossum) [inline README.rst]: they don't want it at SF level, so if this fixer ever makes it into the project it should be at contrib level

Comment (keradus) [inline README.rst]: A bit funny - whole codebase follows that, but we don't want to follow that ;)
After said that, let it be contrib indeed.

Comment (keradus) [inline README.rst]: maybe add it to our `.php_cs` file ?

**Comment by SpacePossum** [comment]
hmmm ping @keradus I think we've to come to an agreement soon about ;
https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1655

tests are now failing even after running the fixer (locally);
https://travis-ci.org/FriendsOfPHP/PHP-CS-Fixer/jobs/103326131

**Comment by SpacePossum** [comment]
I think I got everything to finish this pretty soon. What should we call this Fixer?
`NoConsecutiveUnsetCallsFixer` ?

File: README.rst (hunk 7 above)
Comment (SpacePossum) [inline README.rst]: :+1:  done :)

**Comment by keradus** [comment]
maybe sth like `CombineConsecutiveUnsetsFixer` ?

**Comment by gharlan** [comment]
please add some test cases for `unset($arr['key'])` and `unset($obj->foo)` etc.

**Comment by SpacePossum** [comment]
>please add some test cases for unset($arr['key']) and unset($obj->foo) etc.

:+1:, done, please see:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1662/files#diff-46c9f36495af0a2ab407f69878c2878fR114

I'm out of ideas so please let me know if you a case in mind that is not covered :)

**Comment by gharlan** [comment]
Thanks, that's fine!

:+1:

**Comment by SpacePossum** [comment]
I broke this one as well :(

**Comment by SpacePossum** [comment]
TODO;
run after duplicate semicolon / empty statement fixer

**Comment by SpacePossum** [comment]
the failing test on HHVM are not related with to this PR; https://travis-ci.org/FriendsOfPHP/PHP-CS-Fixer/jobs/109158401

**Comment by keradus** [comment]
Last build on 1.12 was successful:
https://travis-ci.org/FriendsOfPHP/PHP-CS-Fixer/builds/109182719

File: Symfony/CS/Tests/FixerTest.php [hunk 9]
```diff
@@ -215,6 +215,12 @@ public function getFixersPriorityCases()
             array($fixers['no_useless_return'], $fixers['whitespacy_lines']), // tested also in: no_useless_return,whitespacy_lines.test
             array($fixers['no_useless_return'], $fixers['extra_empty_lines']), // tested also in: no_useless_return,extra_empty_lines.test
             array($fixers['no_useless_return'], $fixers['return']), // tested also in: no_useless_return,return.test
+            array($fixers['combine_consecutive_unsets'], $fixers['spaces_after_semicolon']), // tested also in: combine_consecutive_unsets,spaces_after_semicolon.test
+            array($fixers['combine_consecutive_unsets'], $fixers['whitespacy_lines']), // tested also in: combine_consecutive_unsets,whitespacy_lines.test
+            array($fixers['combine_consecutive_unsets'], $fixers['trailing_spaces']), // tested also in: combine_consecutive_unsets,trailing_spaces.test
+            array($fixers['combine_consecutive_unsets'], $fixers['extra_empty_lines']), // tested also in: combine_consecutive_unsets,extra_empty_lines.test
+            array($fixers['duplicate_semicolon'], $fixers['combine_consecutive_unsets']), // tested also in: duplicate_semicolon,combine_consecutive_unsets.test
+
```
Comment (keradus) [inline Symfony/CS/Tests/FixerTest.php]: extra line

File: Symfony/CS/Fixer/Contrib/CombineConsecutiveUnsetsFixer.php [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function fix(\SplFileInfo $file, $content)
+    {
+        $tokens = Tokens::fromCode($content);
+
+        // Offset start at 5, because any call to 'unset' that can possible be merged is
+        // build up out of at least 4 Tokens; 'unset', '(', '[var]', ')' after at least one ';'.
+        for ($index = $tokens->count() - 5; $index > 5; --$index) {
```
Comment (SpacePossum) [inline Symfony/CS/Fixer/Contrib/CombineConsecutiveUnsetsFixer.php]: todo remove

**Comment by keradus** [comment]
Thanks @SpacePossum

**Pull request closed without merging**